### PR TITLE
Python 3.8 and pybind11

### PR DESCRIPTION
**Pull request by ganik**
Remove boost_python dependency
Use pybind11 generated header instead
Upgrade to run with Python3.8
Removed python2.7 & 3.5 support\
Upgrade to 1.8.0 version

### CODE REVIEW

File: src/python/nimbusml/tests/data_type/test_numeric.py [hunk 1]
```diff
@@ -166,4 +166,5 @@ def test_check_series(self):
 
 
 if __name__ == '__main__':
+    TestNumericDataType().test_check_datatype_single_sparse_list_sparse()
```
Comment (xadupre) [inline src/python/nimbusml/tests/data_type/test_numeric.py:169]: I think you can remove this one.

File: src/NativeBridge/DataViewInterop.h [hunk 2]
```diff
@@ -17,7 +17,12 @@ typedef MANAGED_CALLBACK_PTR(bool, GETLABELS)(DataSourceBlock *source, int col,
 // REVIEW: boost_python is not updated at the same speed as swig or pybind11.
 // Both have a larger audience now, see about pybind11 https://github.com/davisking/dlib/issues/293
 // It handles csr_matrix: https://pybind11-rtdtest.readthedocs.io/en/stable/advanced.html#transparent-conversion-of-dense-and-sparse-eigen-data-types.
-using namespace boost::python;
+using namespace pybind11;
+#if !defined(extract_or_cast)
```
Comment (xadupre) [inline src/NativeBridge/DataViewInterop.h:21]: If boost_python is removed, you can probably replace extract_or_cast and has_key_or_contains everywhere.

File: nimbusml.sln [hunk 3]
```diff
@@ -24,142 +24,121 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ManifestGenerator", "src\Ma
...
+		{D3AED287-722F-4243-966E-77AD0652B38E}.DbgLinPy3.8|x64.Build.0 = DbgLinPy3.7|x64
 		{D3AED287-722F-4243-966E-77AD0652B38E}.DbgWinPy3.6|x64.ActiveCfg = DbgWinPy3.6|x64
 		{D3AED287-722F-4243-966E-77AD0652B38E}.DbgWinPy3.6|x64.Build.0 = DbgWinPy3.6|x64
 		{D3AED287-722F-4243-966E-77AD0652B38E}.DbgWinPy3.7|x64.ActiveCfg = DbgWinPy3.7|x64
 		{D3AED287-722F-4243-966E-77AD0652B38E}.DbgWinPy3.7|x64.Build.0 = DbgWinPy3.7|x64
-		{D3AED287-722F-4243-966E-77AD0652B38E}.RlsLinPy2.7|x64.ActiveCfg = RlsLinPy2.7|x64
-		{D3AED287-722F-4243-966E-77AD0652B38E}.RlsLinPy2.7|x64.Build.0 = RlsLinPy2.7|x64
-		{D3AED287-722F-4243-966E-77AD0652B38E}.RlsLinPy3.5|x64.ActiveCfg = RlsLinPy3.5|x64
-		{D3AED287-722F-4243-966E-77AD0652B38E}.RlsLinPy3.5|x64.Build.0 = RlsLinPy3.5|x64
+		{D3AED287-722F-4243-966E-77AD0652B38E}.DbgWinPy3.8|x64.ActiveCfg = DbgWinPy3.7|x64
```
Comment (xadupre) [inline nimbusml.sln:122]: I wonder if it is an issue to have different version on the same line.

File: build.cmd [hunk 4]
```diff
@@ -188,7 +152,7 @@ echo "Building Managed code ... "
 echo "#################################"
 set _dotnet=%_dotnetRoot%\dotnet.exe
 
-if "%Configuration:~-5%" == "Py3.7" set VerifyManifest=True
+if "%Configuration:~-5%" == "Py3.6" set VerifyManifest=True
```
Comment (pieths) [inline build.cmd:155]: Why was this version lowered vs. increased since the default configuration is now Python 3.8?

File: build.cmd [hunk 5]
```diff
@@ -393,14 +349,10 @@ if "%InstallPythonPackages%" == "True" (
     echo "Installing python packages ... "
     echo "#################################"
     call "%PythonExe%" -m pip install --upgrade "pip==19.3.1"
```
Comment (pieths) [inline build.cmd:351]: I remember that there were a few issues with newer versions of `pip`. Are these issues still there or can the version of `pip` be upgraded to the latest?

File: src/NativeBridge/stdafx.h [hunk 6]
```diff
@@ -111,5 +107,14 @@ class StopWatch
 //
 // frequently used namespace aliases
 //
-namespace bp = boost::python;
-namespace np = boost::python::numpy;
\ No newline at end of file
+namespace bp = pybind11;
```
Comment (pieths) [inline src/NativeBridge/stdafx.h:110]: Should `bp` be switched with `pb` here and throughout the rest of the code base to fully remove any references to boost python?

(Or switched to `py` which is used in some of the pybind examples. Though `py` seems to be a little too similar to the standard native python naming scheme)

File: src/python/nimbusml/tests/scikit/test_uci_adult_scikit.py [hunk 7]
```diff
@@ -334,4 +335,5 @@ def test_lr_named_steps_iris(self):
 
 
 if __name__ == '__main__':
+    TestUciAdultScikit().test_feature_union()
```
Comment (pieths) [inline src/python/nimbusml/tests/scikit/test_uci_adult_scikit.py:338]: Wouldn't this be run automatically when the file is run standalone?

File: build.cmd (hunk 4 above)
Comment (ganik) [reply build.cmd:155]: this is non issue. it should work both on 3.7 or 3.6

---
In reply to: [441269724](https://github.com/microsoft/NimbusML/pull/494#discussion_r441269724) [](ancestors = 441269724)

File: build.cmd (hunk 5 above)
Comment (ganik) [reply build.cmd:351]: not relevant to this PR. will address it separately

---
In reply to: [441272014](https://github.com/microsoft/NimbusML/pull/494#discussion_r441272014) [](ancestors = 441272014)

**Comment by ganik** [comment]
this is related to python installation I use in testing. Should not happen with proper python installation

---
In reply to: [432877408](https://github.com/microsoft/NimbusML/pull/494#issuecomment-432877408) [](ancestors = 432877408)

**Pull request merged**